gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java

```diff
@@ -2084,8 +2084,13 @@ public boolean updateUngradedItems(final long assignmentId, final double grade,
 
 				// TODO if this is slow doing it one by one, might be able to
 				// batch it
+
+				// The service needs it otherwise it will assume 'null'
+				// so pull it back from the service and poke it in there!
+				final String comment = getAssignmentGradeComment(Long.valueOf(assignmentId), studentUuid);
+
 				this.gradebookService.saveGradeAndCommentForStudent(gradebook.getUid(), assignmentId, studentUuid,
```

Comment:
I think you can just call `gradebookService.saveGrade()`, rather than `saveGradeAndCommentForStudent`. In this way, the comments will be left untouched, and you don't have to query the database for the original comment to 're-save' it.

It will probably improve the performance of the algorithm as a whole as well.

Reply:
Hmm, I looked for saveGrade in edu-services and didn't see it. 

I am seeing it in the GradebookNgBusinessService.java (local to gradebookng)

I looked through the code there and there is a comment
// note, you must pass in the comment or it will be nulled out by the GB service

So it doesn't look like that method would work.

The problem is that it needs a way to be able to clear out the grade, so null is the way to go. It could have a new completely separate method saveGradeWithoutComment or just saveGradeForStudent (or something?) but it seems like it would reproduce a lot of logic for saveGradeAndCommentForStudent. I couldn't think of any quick way to easily combine them without 100 new lines of code. (Which I didn't have time to do.

If someone wants to take a pass at that go for it, but this also had the least risk when we're close to release. If you look through the saveGradeAndCommentForStudent code it does a lot of selects and queries that could also be further optimized, and this method could as well. I only tested on a course of 200 students.

Reply:
Are you seeing saveGrade in your service? Maybe you have a local modification that you can provide?

Reply:
Sorry, when I said `gradebookService.saveGrade()`, I should have said 'GradebookNgBusinessService.saveGrade()`, but now that I look at the implementation of that method, I see what you're saying. It's just calling the same edu-services method behind the scenes and you'd have the same problem with the comment getting nulled out.

I supposed there is the opportunity to create a new method in edu-services to do just the grade saving, but I'm not sure if you want to take on that extra work.